[17.0][IMP] l10nl_bank script

bosd opened the pull request:
As mentioned in https://github.com/OCA/l10n-netherlands/pull/424

bosd committed:
[IMP] l10n_nl_bank: add Generate datafile script.

bosd committed:
[IMP] l10n_nl_bank: re-generate data file.

StefanRijnhart commented on l10n_nl_bank/data/res_bank_data.xml:
```diff
@@ -140,174 +158,169 @@
+    </record>
     <record id="INGB" model="res.bank">
-        <field name="name">ING Bank N.V.</field>
+        <field name="name">Ing</field>
```
I see you took special care to try and keep the casing, but as a result banks like 'ASN' and 'ING' are now assigned incorrect case. Maybe we should stick to the capitalization as in the file (e.g. 'ASN', 'ING' but also 'BANK NEDERLANDSE GEMEENTEN')?

bosd replied:
Quickly edited the override list. Now we have nice casing :wink:

StefanRijnhart commented on l10n_nl_bank/readme/CONTRIBUTORS.md:
```diff
@@ -1,2 +1,3 @@
 - Erwin van der Ploeg \<<erwin@odooexperts.nl>\>
 - Andrea Stirpe \<<a.stirpe@onestein.nl>\>
+- bosd
```
Please add at least your first and last name if you don't want your email in here.

hbrunn commented on l10n_nl_bank/scripts/generate_bank_data.py:
```diff
@@ -0,0 +1,175 @@
+            processed_banks.append((record_id, name, bic))
+
+        logging.info("✅ Data processed successfully.")
+        return sorted(processed_banks, key=lambda x: x[1])
```
```suggestion
        return sorted(processed_banks, key=lambda x: x[0])
```

to keep sorting by bic as the existing file does, should give a smaller diff and keep ordering stable for name changes

hbrunn reviewed (changes requested):
we'll need a migration script here that renames the xmlid of adyen and possibly other ones where the xmlid is changed. and then force-loads the file to update names like sns.

if you can easily instruct your statistical parrot to rewrite the script to use xlrd and requests instead of the pretty heavy pandas dependency, it would be nice because the former are already dependencies of odoo. but i won't block on this one

bosd commented:
@hbrunn Thanks for the review.
:parrot: Migration scripts added.
moved to openpyxl so from version 18 and on we are back in sync. (xlrd no longer supports xlsx files)

hbrunn commented on l10n_nl_bank/migrations/17.0.1.1.0/post-migration.py:
```diff
@@ -0,0 +1,38 @@
+
+
+def migrate(cr, version):
+    """
+    This post-migration script re-enables the noupdate="1" flag on the bank
+    data file to prevent accidental data overwrites during future module upgrades.
+    """
```
the comments in this file are very misleading, as they don't match what the code actually does

hbrunn commented on l10n_nl_bank/migrations/17.0.1.1.0/post-migration.py:
```diff
@@ -0,0 +1,38 @@
+    This post-migration script re-enables the noupdate="1" flag on the bank
+    data file to prevent accidental data overwrites during future module upgrades.
+    """
+    _logger.info("Starting post-migration: Restoring noupdate flag.")
```
it's quite uncommon to be that verbose, but then it also doesn't really hurt

hbrunn commented on l10n_nl_bank/migrations/17.0.1.1.0/post-migration.py:
```diff
@@ -0,0 +1,38 @@
+    data file to prevent accidental data overwrites during future module upgrades.
+    """
+    _logger.info("Starting post-migration: Restoring noupdate flag.")
+    api.Environment(cr, SUPERUSER_ID, {})
```
this statement does nothing

hbrunn commented on l10n_nl_bank/migrations/17.0.1.1.0/post-migration.py:
```diff
@@ -0,0 +1,38 @@
+        _logger.info(
+            "Post-migration successful: Restored noupdate='1' for %s.", xml_file_path
+        )
+    except Exception as e:
```
generally, if some step of a migration script fails, the whole thing should fail, so the exception handling shouldn't be there at all.

if this code ever ran it would have also hidden all the things wrong with it

hbrunn commented on l10n_nl_bank/migrations/17.0.1.1.0/post-migration.py:
```diff
@@ -0,0 +1,38 @@
+
+    # Define the XML file to be reloaded. This is the same file that was temporarily
+    # changed in the pre-migration step.
+    xml_file_path = "l10n_nl_bank/data/res_bank_data.xml"
```
it's wrong to pass the path of the module

hbrunn commented on l10n_nl_bank/migrations/17.0.1.1.0/pre-migration.py:
```diff
@@ -0,0 +1,51 @@
+    old_xml_id = "ADYEN"
+    new_xml_id = "ADYB"
+
+    # Get the module name. This is crucial for a multi-module environment.
+    try:
+        module = env.ref("l10n_nl_bank").module
+    except ValueError:
+        _logger.error(
+            "Could not determine module name. Please check if 'l10n_nl_bank' exists."
+        )
+        return
```
this whole section is nonsense and causes the script to exit early

hbrunn commented on l10n_nl_bank/migrations/17.0.1.1.0/pre-migration.py:
```diff
@@ -0,0 +1,51 @@
+
+
+def migrate(cr, version):
+    """
+    This pre-migration script updates the XML ID for the Adyen bank
+    and temporarily sets noupdate="0" on the bank data file
+    to allow for a full data reload during the module upgrade.
+    """
```
none of the above is true

hbrunn commented on l10n_nl_bank/migrations/17.0.1.1.0/post-migration.py:
```diff
@@ -0,0 +1,38 @@
+            mode="update",
+            noupdate=True,
+            kind="data",
+        )
```
this call does not actually cause data to be overwritten, and fails anyways

hbrunn reviewed (changes requested):
nearly every line of the migration scripts is wrong, but they won't run anyways because the version number isn't increased in the manifest

please only submit code you tested and verified to work as expected

bosd commented:
@hbrunn thanks, you were right on all counts — the migration scripts were broken and weren't even running because the manifest version wasn't bumped. Reworked both:

- **`__manifest__.py`**: bumped to `17.0.1.1.0` so the migration directory actually runs.
- **pre-migration**: dropped the bogus "get module name" block (which exited early) and the misleading docstring. The only genuine xmlid rename is `ADYEN` → `ADYB` (same BIC, the id is now derived from the BIC prefix like every other record), so it's just one `openupgrade.rename_xmlids` call.
- **post-migration**: the old `convert_file(..., noupdate=True)` didn't overwrite anything (and failed). The data file is `noupdate="1"`, so existing names aren't refreshed on a normal upgrade. Now it reloads via `openupgrade.load_data(..., mode="init")`, which bypasses the noupdate skip so existing names (e.g. SNS) get updated and the new banks get created. Also fixed the module-relative path.

Both now use the `@openupgrade.migrate()` helper and dropped the inappropriate exception handling / no-op `Environment` call. Verified pre-commit is green and folded the fixes into the original commit.
